### ISSUE TITLE
Grievances with the tree-sitter query language

### ISSUE THREAD

**Issue by nbacquey** [tweag/topiary#276]
This issue will only serve to discuss weaknesses of the tree-sitter query language. We will be able to address them separately if needed.

### We can't use the (apparent?) full set of tree-sitter queries features
The [documentation](https://tree-sitter.github.io/tree-sitter/using-parsers#predicates) states that we can use predicates of the form
```scheme
(
  (identifier) @constant
  (#match? @constant "^[A-Z][A-Z_]+")
)
```
However, we can't currently compile them in the Rust library.

TODO: develop each point and give examples

**Comment by Xophmeister** [comment]
Great list :+1: Just to expand a little, from my experience:

* Negation is amongst the Tree-Sitter issues (https://github.com/tree-sitter/tree-sitter/issues/705), but has been dormant for 2+ years.
* Anecdotally, the Kleene star appears to ignore anonymous nodes. (That mightn't be the exact constraint; more data needed!)
* At a glance, the `#match?` and `#eq?` predicates feel like they could be useful. However, I wonder how Topiary would use them, as we're using capture names as formatting directives, rather than arbitrary labels.

**Comment by Xophmeister** [comment]
Another one:
* I have found myself needing, a few times, to write queries of the form:

  ```scm
  (X
    Y
    ; etc.
  )
  ```

  Where `X` ranges over a few nodes (say, `a`, `b` and `c`) and `Y` is constant. It would be nice if it were possible to alternate over head nodes (e.g., `([a b c] Y)`), rather than rewrite the rule several times and/or using scopes.

**Comment by nbacquey** [comment]
Updated the initial message, and added your use-case @Xophmeister

**Comment by nbacquey** [comment]
Added another one. I'm thinking about committing this file to the repository itself, what do you think?

**Comment by aspiwack** [comment]
I don't see why not. But on the other hand, what purpose would serve a file that gathers this information?

**Comment by nbacquey** [comment]
The same as this issue, I think:
* Warn contributors about the known shortcomings of the query language.
* Give pointers on which contributions to `tree-sitter` would be appreciated.

Except that it would stay visible, and its modifications would be `git`-tracked, which makes following its evolution easier than tracking a particular issue.

**Comment by nbacquey** [comment]
I suppose another way to achieve that would be to add a `query language` issue tag, and create one issue per point

**Comment by Xophmeister** [comment]
This is more of a meta-issue, but I figured it's worth mentioning here as it's somewhat relevant:

Our language query files are declarative formatting directives, where each query targets some syntactic structure attested by the grammar. Targets can overlap and, in general (AFAIK) there is no limit to the number of queries that can target a specific node. For non-trivial languages, this quickly flaunts the usual good practices of writing code. For example, language query files can become hundreds-to-thousands of lines long and it's up to whoever wrote them to organise them coherently and remember the state, such that any new queries don't conflict or cause unintended interactions.

By in large, this is doable rather than intractable, because programming languages' syntactic structures are (anecdotally) relatively well siloed. However, edge cases certainly exist where it's not immediately obvious which queries are being applied.

**Comment by aspiwack** [comment]
@nbacquey I think it depends on the objective (or maybe we need both). If the goal is to discuss how they can be addressed, to create workarounds in our workflows or upstream contributions, then one-issue-per-grievance is the way to go. If the goal is to document gotchas when writing query files, then this should be part of our documentation.

Maybe we need a bit of both.

@Xophmeister I don't think that there is a fundamental solution to this. In that there probably isn't an ideal formatting DSL that doesn't exhibit the problem. Maybe what we could think about is to have debugging tools that make clear which queries have matched a particular piece of code?

**Comment by Xophmeister** [comment]
> I don't think that there is a fundamental solution to this. In that there probably isn't an ideal formatting DSL that doesn't exhibit the problem. Maybe what we could think about is to have debugging tools that make clear which queries have matched a particular piece of code?

Indeed; some kind of tooling to aid development is what I had in mind :+1:

**Comment by nbacquey** [comment]
I was thinking about a debugging tool, or a debugging mode, as well.

**Comment by torhovland** [comment]
One more point:

There is no way to define constants to reduce duplication. For example, the Bash grammar uses this particular construct 9 times:

```scheme
[(command) (list) (pipeline) (compound_statement) (subshell) (redirected_statement) (variable_assignment)]
```

**Comment by Xophmeister** [comment]
> There is no way to define constants to reduce duplication.

A metalanguage that compiles down to Tree-sitter queries _sounds_ quite attractive. (Leveraging Nickel to do the job would be the icing on the cake!)

**Comment by torhovland** [comment]
Good point. And actually this particular problem (reusable constants) could be solved with the most basic templating solution imaginable.

**Comment by aspiwack** [comment]
@nbacquey I was reading some documentation, and, for completeness, there is a limited form of negation in the query language

> Negated Fields
>
> You can also constrain a pattern so that it only matches nodes that lack a certain field. To do this, add a field name prefixed by a ! within the parent pattern. For example, this pattern would match a class declaration with no type parameters:
>
> ```
> (class_declaration
>   name: (identifier) @class_name
>   !type_parameters)
> ```

**Comment by aspiwack** [comment]
> However, we can't currently compile them in the Rust library.

The documentation explicitly calls out `#match?` and `#eq?` as being understood by the Rust crate (as far as tree-sitter proper is concerned, both are supposed to be just understood as uninterpreted predicates, but the Rust crate is supposed to impose semantics on top of them). So it would suggest that maybe we're calling the API wrong, or that there is a bug in the Rust crate implementation.

**Comment by aspiwack** [comment]
> The anchor . ignores non-named nodes, except when it doesn't

I haven't seen an example of unnamed nodes not being ignored here. Is this grievance's title misleading?

I think that it's clear from the `(comment)` issue that unnamed nodes are not quite what you would need to skip, though.

Your example does document a pretty important pitfall though: it's typically a bad idea to use `_` as an operand of `.`

**Comment by aspiwack** [comment]
> The Kleene star * can match non-consecutive sequences of nodes

I'm not sure that your example demonstrates this. It looks like it may be an occurrence of the somewhat counter intuitive behaviour of `.` discussed above (both above in your issue, and above as in the previous comment :slightly_smiling_face: ).

Maybe we should try a more constrained query

```
(
  "["
  .
  (number)
  .
  (number)*
  .
  (number) @append_delimiter
  .
  "]"
  (#delimiter! "#")
)`
```

And run it on various length of lists and see which one match (if the `*` is well-behaved, I expect this query to match lists of size 2 and of size 3. But not of size 4. If `(number)*` matches non-consecutive number, then size 4 and above should be matched.

**Comment by torhovland** [comment]
> > However, we can't currently compile them in the Rust library.
> 
> The documentation explicitly calls out `#match?` and `#eq?` as being understood by the Rust crate (as far as tree-sitter proper is concerned, both are supposed to be just understood as uninterpreted predicates, but the Rust crate is supposed to impose semantics on top of them). So it would suggest that maybe we're calling the API wrong, or that there is a bug in the Rust crate implementation.

Can't remember the exact details now, but I do remember not being able to use built-in predicates, and finding that they were indeed not implemented in the Rust binding. Worth another look.

**Comment by Xophmeister** [comment]
> [snip] there is a limited form of negation in the query language
> 
> > Negated Fields
> > You can also constrain a pattern so that it only matches nodes that lack a certain field. To do this, add a field name prefixed by a ! within the parent pattern. For example, this pattern would match a class declaration with no type parameters:
> > ```
> > (class_declaration
> >   name: (identifier) @class_name
> >   !type_parameters)
> > ```

AFAIK, the need for this kind of negation has never come up. It arises in a more general sense, when there's a pattern that applies to a strict subset of what the grammar affords (i.e., you end up with "exceptional nodes") and `@do_nothing` can't be applied. So far, the only solution we've found is to replicate that pattern for every node in that subset.

**Comment by nbacquey** [comment]
Added a paragraph on tokens captured by regular expressions in the grammar

**Comment by 314eter** [comment]
I never realised the difference between strings and regular expressions in a grammar until tree-sitter/tree-sitter-ocaml#63. I created tree-sitter/tree-sitter-ocaml#77 to make the behaviour for operators more consistent. @nbacquey do you think that's a good idea, or will it make it even more complicated for topiary?

**Comment by 314eter** [comment]
Done in [v0.20.2](https://github.com/tree-sitter/tree-sitter-ocaml/releases/tag/v0.20.2).

**Comment by torhovland** [comment]
Also see https://github.com/tweag/topiary/pull/537#issuecomment-1609166379
